File: CJUIKitDemo/CJUIKitDemo/CommonUI/Base/CJUIKitBaseScrollViewController.m
//
//  CJUIKitBaseScrollViewController.m
//  CJUIKitDemo
//
//  Created by ciyouzen on 2015/12/23.
//  Copyright © 2015年 dvlproad. All rights reserved.
//

#import "CJUIKitBaseScrollViewController.h"

@interface CJUIKitBaseScrollViewController ()

@end

@implementation CJUIKitBaseScrollViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    
    UIScrollView *scrollView = [[UIScrollView alloc] init];
    scrollView.backgroundColor = CJColorFromHexString(@"#f2f2f2");
    [self.view addSubview:scrollView];
    [scrollView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.right.top.bottom.mas_equalTo(self.view);
    }];
    self.scrollView = scrollView;
    
    UIView *containerView = [[UIView alloc] init];
    //containerView.backgroundColor = [UIColor greenColor];
    [scrollView addSubview:containerView];
    [containerView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.right.mas_equalTo(scrollView);
        make.top.bottom.mas_equalTo(scrollView);
        make.width.mas_equalTo(scrollView.mas_width);
        make.height.mas_equalTo(scrollView.mas_height).mas_offset(1);
    }];
    self.containerView = containerView;
}

/// 更新containerView与lastBottomView的底部间隔，未调用时containerView的高为比scrollView多1像素
- (void)updateScrollHeightWithLastBottomView:(UIView *)lastBottomView bottom:(CGFloat)bottom {
    NSAssert(bottom > 0, @"Error:scrollView与lastBottomView的底部间隔必须大于0");
    
    [self.containerView mas_remakeConstraints:^(MASConstraintMaker *make) {
        make.left.right.mas_equalTo(self.scrollView);
        make.top.bottom.mas_equalTo(self.scrollView);
        make.width.mas_equalTo(self.scrollView.mas_width);
        make.bottom.mas_equalTo(lastBottomView.mas_bottom).mas_offset(bottom);
    }];
}

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

@end

File: CJUIKitSwiftDemo/CJUIKitSwiftDemo/CommonUI/Base/CJUIKitBaseScrollViewController.swift
//
//  CJUIKitBaseScrollViewController.m
//  CJUIKitDemo
//
//  Created by ciyouzen on 2015/12/23.
//  Copyright © 2015年 dvlproad. All rights reserved.
//

import SnapKit

class CJUIKitBaseScrollViewController: CJUIKitBaseViewController {
    fileprivate var scrollView: UIScrollView!
    fileprivate var containerView: UIView!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view.
        let scrollView: UIScrollView = UIScrollView()
        scrollView.backgroundColor = CJColorFromHexString("#f2f2f2")
        self.view.addSubview(scrollView)
        scrollView.snp.makeConstraints { (make) in
            make.edges.equalToSuperview()
        }
        self.scrollView = scrollView
        
        
        
        let containerView: UIView = UIView()
        containerView.backgroundColor = UIColor.green;
        scrollView.addSubview(containerView)
        containerView.snp.makeConstraints { (make) in
            make.left.right.equalTo(scrollView)
            make.top.bottom.equalTo(scrollView)
            make.width.equalTo(scrollView.snp_width)
            make.height.equalTo(scrollView.snp_height).offset(1)
        }
        self.containerView = containerView
    }
    
    
    /**
     *  更新ScrollView的高
     *  @brief  ①如果没有lastBottomView来确认scrollView的高，那么高为根据父视图设置；
     ②如果有lastBottomView，则通过设置scrollView的containerView与lastBottomView的底部间隔来更新ScrollView的高
     *
     *  @param bottomInterval bottomInterval
     *  @param lastBottomView lastBottomView(可为nil)
     */
    func updateScrollHeight(bottomInterval:CGFloat, lastBottomView:UIView?) {
        assert(bottomInterval >= 0, "Error:scrollView与lastBottomView的底部间隔不能小于0");
        
        if lastBottomView == nil {
            self.containerView.snp.remakeConstraints { (make) in
                make.left.right.equalTo(self.scrollView);
                make.top.bottom.equalTo(self.scrollView);
                make.width.equalTo(self.scrollView.snp_width);
                make.height.equalTo(self.scrollView.snp_height).offset(bottomInterval);
            }
        } else {
            self.containerView.snp.remakeConstraints { (make) in
                make.left.right.equalTo(self.scrollView);
                make.top.bottom.equalTo(self.scrollView);
                make.width.equalTo(self.scrollView.snp_width);
                make.bottom.equalTo(lastBottomView!.snp_bottom).offset(bottomInterval);
            }
        }
    }
    
}
